File: Python/15.三数之和.py
from typing import List

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if len(nums) < 3:
            return []
        ans: List[List[int]] = []
        nums.sort()
        for i in range(0, len(nums) - 2):
            target = -nums[i]
            left, right = i + 1, len(nums) - 1
            while left < right:
                if nums[left] + nums[right] == target:
                    if [-target, nums[left], nums[right]] not in ans:
                        ans.append([-target, nums[left], nums[right]])
                    left += 1
                    right -= 1
                elif nums[left] + nums[right] < target:
                    left += 1
                else:
                    right -= 1
        return ans

File: Java/15.三数之和.java
package Java;

import java.util.*;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        if (nums.length < 3) {
            return Arrays.asList();
        }
        Arrays.sort(nums);
        Set<String> st = new HashSet<String>();
        List<List<Integer>> ans = new ArrayList<List<Integer>>();
        for (int i = 0; i < nums.length - 2; ++i) {
            int target = -nums[i];
            int left = i+1, right = nums.length - 1;
            while (left < right) {
                if (nums[left] + nums[right] == target) {
                    List<Integer> tmpl = Arrays.asList(-target, nums[left], nums[right]);
                    String tmps = tmpl.toString();
                    if (!st.contains(tmps)) {
                        st.add(tmps);
                        ans.add(tmpl);
                    }
                    ++left;
                    --right;
                } else if (nums[left] + nums[right] < target) {
                    ++left;
                } else {
                    --right;
                }
            }
        }
        return ans;
    }
}
